lib/Dialect/AIEX/Transforms/AIEMaterializeBDChains.cpp

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc.
+//
+//===----------------------------------------------------------------------===//
+
+#include <set>
+
+#include "aie/Dialect/AIE/IR/AIEDialect.h"
+#include "aie/Dialect/AIEX/IR/AIEXDialect.h"
+#include "aie/Dialect/AIEX/Transforms/AIEXPasses.h"
+
+#include "mlir/Analysis/CallGraph.h"
+#include "mlir/Interfaces/CallInterfaces.h"
+#include "mlir/Pass/AnalysisManager.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Inliner.h"
+
+using namespace mlir;
+using namespace xilinx;
+using namespace xilinx::AIEX;
+
+struct AIEMaterializeBDChainsPass
+    : AIEMaterializeBDChainsBase<AIEMaterializeBDChainsPass> {
+
+  WalkResult inlineUsage(AIE::DeviceOp device, DMAStartBdChainOp start_op) {
+    OpBuilder builder = OpBuilder(start_op);
+
+    // Get referenced abstract BD chain
+    AIE::BDChainOp chain_def = start_op.getBDChainOp();
+    assert(chain_def);
+    Region &source_region = chain_def.getBody();
+
+    // Create BD op into which the result will be inlined
+    DMAConfigureTaskOp configure_op = builder.create<DMAConfigureTaskOp>(
+        start_op.getLoc(), builder.getIndexType(), start_op.getTile(),
+        start_op.getDirection(), start_op.getChannel(),
+        start_op.getIssueToken(), start_op.getRepeatCount());
+    Region &target_region = configure_op.getBody();
+
+    // Clone BD definition into usage site, replacing abstract SSA values with
+    // concrete ones
+    IRMapping arg_map;
+    ValueRange values = start_op.getConcreteArgs();
+    for (unsigned i = 0, n = source_region.getNumArguments(); i < n; i++) {
+      BlockArgument arg = source_region.getArgument(i);
+      Value val = values[i];
+      assert(arg.getType() == val.getType());
+      arg_map.map(arg, val);
+    }
+    source_region.cloneInto(&target_region, arg_map);
+
+    // Replace result of dma start task with result of bd chain configuration
+    start_op.getResult().replaceAllUsesWith(configure_op.getResult());
+
+    // Remove definition too if this was the only/last usage of it
+    if (SymbolTable::symbolKnownUseEmpty(chain_def, device)) {
+      chain_def.erase();
+    }
+
+    // Add a start BDs instruction
+    builder.create<DMAStartTaskOp>(start_op.getLoc(), configure_op.getResult());
+
+    // After fully inlining, remove the original instruction
+    start_op.erase();
+
+    return WalkResult::advance();
+  }
+
+  void runOnOperation() override {
+    WalkResult r;
+
+    AIE::DeviceOp device = getOperation();
+
+    // Wrap bd chains in DMAConfigureTaskOp regions before inlining
+    r = device.walk([&](DMAStartBdChainOp start_op) {
+      return inlineUsage(device, start_op);
+    });
+    if (r.wasInterrupted()) {
+      return signalPassFailure();
+    }
+
+    // Verify inlined basic blocks do form a chain reachable from the start;
+    // Remove empty blocks
```

Comment:
Can "remove empty blocks" be a canonicalization pattern?

Reply:
Do canonicalization patterns run after each pass? It would be important that empty blocks are removed before subsequent lowering passes, because it may trip `--aie-dma-tasks-to-npu` up if it finds a basic block containing no `dma_bd` ops.

Reply:
It turns out we don't have any examples of op canonicalization in mlir-aie (yet).

If you put let `hasCanonicalizer = 1;` in the tablegen description for an op, then you get to define `op::getCanonicalizationPatterns` in your Dialect.cpp, and in our case your `runOnOperation` can be rewritten like this:
```c++
    RewritePatternSet patterns(ctx);
    patterns.insert<MaterializeBDChainsPattern>(ctx);
    DMAConfigureTaskOp::getCanonicalizationPatterns(patterns, ctx);
    (void)applyPatternsAndFoldGreedily(device, std::move(patterns));
```
with the benefit that whatever patterns you put in `op::getCanonicalizationPatterns` will also be used by `-canonicalize`